[PGNCCL] Fix bugs in non-blocking mode

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #137544
* #137855
* __->__ #137741

### Fix 1: Throw async error during init wait

Previously we just busy wait for `ncclSuccess`, if the nonblocking init encountered error, we never report that. Added detection of async error via `ncclGetAsyncError`.

### Fix 2: Add wait after comm split

```
  // After calling ncclCommSplit in non-blocking mode, we should wait for the
  // source communicator to be out of ncclInProgress state.
  // Reason 1:
  //   it's unsafe to call new operations on the parent comm while it's in
  //   ncclInProgress state.
  // Reason 2:
  //   as of NCCL 2.23, the ptr value of child comm will not be filled until the
  //   state of parent comm is ncclSuccess. This may change in the future. See:
  //   https://github.com/NVIDIA/nccl/issues/1472
```
This wait does not mean the child comm is ready for use, neither does it block till that point.

cc @XilunWu @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

kwen2501 committed:
[PGNCCL] Fix bugs in non-blocking mode

- Throw async error during init wait

- Add wait after comm split

[ghstack-poisoned]

kwen2501 committed:
Update on "[PGNCCL] Fix bugs in non-blocking mode"


- Throw async error during init wait

- Add wait after comm split

cc XilunWu H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/NCCLUtils.cpp:
```diff
@@ -80,14 +64,30 @@ std::shared_ptr<NCCLComm> NCCLComm::split(
       ncclCommSplit(sourceComm, color_id, rank, &(comm->ncclComm_), &config),
       std::nullopt);
+#else
+  // After calling ncclCommSplit in non-blocking mode, we should wait for the
```
i'm kinda confused about the meaning of non-blocking mode in nccl.  maybe it would be a good time to write up a note on different modes of nccl init and what their semantics are.

are there 2 states that we transition through in non-blocking init?  1) ncclInProgress state, 2) still doing some more non-blocking init, finally 3) ready for use?  And the point is we need to block/wait on (1) but we can let (2) be asynchronous?  otherwise i'm confused about what part of non-blocking init is actually non-blocking

kwen2501 replied:
> the meaning of non-blocking mode in nccl

In general, it means the API call will return immediately, instead of blocking the calling thread until the init, finalize, etc operations complete.

> are there 2 states that we transition through in non-blocking init? 1) ncclInProgress state, 2) still doing some more non-blocking init, finally 3) ready for use? 

In general, two states only: ncclInProgress and ncclSuccess (if there is no error).

`ncclCommSplit` is a special case: it involves two communicators -- the parent comm and the child comm.
As of NCCL 2.23 (today): when `ncclCommSplit` returns, a valid pointer for child comm may not have been assigned. We'd need to wait for `ncclSuccess` on the parent comm to confirm that. But that doesn't mean the child comm is ready for use, we need to wait for `ncclSuccess` on the child comm to confirm that.

I have talked to the NCCL team re improvement of the expectation, i.e. when `ncclCommSplit` returns, a valid pointer for child comm SHOULD have been assigned. It seems NCCL team agrees, so that the API presents non-blocking semantics for both the parent and the child comms. Detailed discussion is here:
https://github.com/NVIDIA/nccl/issues/1472

shuqiangzhang commented on torch/csrc/distributed/c10d/NCCLUtils.cpp:
```diff
@@ -34,37 +30,25 @@ ncclComm_t NCCLComm::getNcclComm() {
-  // only wait for initialization if nonblocking mode is enabled
-  if (!initialized_ && nccl_use_nonblocking()) {
-    waitUntilInitialized(nccl_nonblocking_timeout());
+  if (!initialized_) {
```
Previous implementation should not hang forever? as there is time bound on timeout?

kwen2501 replied:
It will not hang forever, if it is just regular hang.
But it will hang forever, if there is an async error using init.

eqy replied:
If non-blocking init itself failed, does NCCL know (we are in non-blocking mode?) to populate the error in `ncclGetAsyncError` ? (I'm assuming `ncclGetAsyncError` only works in nonblocking mode.

kwen2501 replied:
`ncclGetAsyncError` will populate error when non-blocking init fails. 

`ncclGetAsyncError` also works in blocking mode, i.e. when checking if a running collective hit any network error.

shuqiangzhang commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -128,6 +132,24 @@
   }
 
+#define C10D_SCHED_SLEEP()     \
+  std::this_thread::sleep_for( \
```
Define an inline function instead of a macro?

kwen2501 replied:
That's true in general. The special case here is that I am doing macro expansion below and passing function pointer does not seem very straightforward.

shuqiangzhang commented on torch/csrc/distributed/c10d/NCCLUtils.cpp:
```diff
@@ -80,14 +64,30 @@ std::shared_ptr<NCCLComm> NCCLComm::split(
+  // source communicator to be out of ncclInProgress state.
+  // Reason 1:
+  //   it's unsafe to call new operations on the parent comm while it's in
+  //   ncclInProgress state.
```
line 66 "auto sourceComm = source->getNcclComm();" should have already guaranteed that the parent comm is ready?

kwen2501 replied:
The wait here refers to *AFTER* calling `ncclCommSplit` -- we need to wait on `sourceComm` to make sure it gives a valid pointer to child comm.

shuqiangzhang replied:
Still confused, before we call ncclCommSplit here, the parent comm should have already been ncclSuccess state instead of ncclInProgress state, because there is already a wait in getNcclComm(). We can discuss this offline as it seems I misunderstood something

kwen2501 replied:
Would this discussion help clarify?
https://github.com/NVIDIA/nccl/issues/1472#issuecomment-2402892251

shuqiangzhang replied:
Then you might need update the comment? because the parent comm is ready/initialized, but it is the child comm which is not initialized, so we need to wait here.

kwen2501 committed:
Update on "[PGNCCL] Fix bugs in non-blocking mode"


### Fix 1: Throw async error during init wait

Previously we just busy wait for `ncclSuccess`, if the nonblocking init encountered error, we never report that. Added detection of async error via `ncclGetAsyncError`.

### Fix 2: Add wait after comm split

```
  // After calling ncclCommSplit in non-blocking mode, we should wait for the
  // source communicator to be out of ncclInProgress state.
  // Reason 1:
  //   it's unsafe to call new operations on the parent comm while it's in
  //   ncclInProgress state.
  // Reason 2:
  //   as of NCCL 2.23, the ptr value of child comm will not be filled until the
  //   state of parent comm is ncclSuccess. This may change in the future. See:
  //   https://github.com/NVIDIA/nccl/issues/1472
```
This wait does not mean the child comm is ready for use, neither does it block till that point.

cc XilunWu H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 committed:
Update on "[PGNCCL] Fix bugs in non-blocking mode"


### Fix 1: Throw async error during init wait

Previously we just busy wait for `ncclSuccess`, if the nonblocking init encountered error, we never report that. Added detection of async error via `ncclGetAsyncError`.

### Fix 2: Add wait after comm split

```
  // After calling ncclCommSplit in non-blocking mode, we should wait for the
  // source communicator to be out of ncclInProgress state.
  // Reason 1:
  //   it's unsafe to call new operations on the parent comm while it's in
  //   ncclInProgress state.
  // Reason 2:
  //   as of NCCL 2.23, the ptr value of child comm will not be filled until the
  //   state of parent comm is ncclSuccess. This may change in the future. See:
  //   https://github.com/NVIDIA/nccl/issues/1472
```
This wait does not mean the child comm is ready for use, neither does it block till that point.

cc XilunWu H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 committed:
Update on "[PGNCCL] Fix bugs in non-blocking mode"


### Fix 1: Throw async error during init wait

Previously we just busy wait for `ncclSuccess`, if the nonblocking init encountered error, we never report that. Added detection of async error via `ncclGetAsyncError`.

### Fix 2: Add wait after comm split

```
  // After calling ncclCommSplit in non-blocking mode, we should wait for the
  // source communicator to be out of ncclInProgress state.
  // Reason 1:
  //   it's unsafe to call new operations on the parent comm while it's in
  //   ncclInProgress state.
  // Reason 2:
  //   as of NCCL 2.23, the ptr value of child comm will not be filled until the
  //   state of parent comm is ncclSuccess. This may change in the future. See:
  //   https://github.com/NVIDIA/nccl/issues/1472
```
This wait does not mean the child comm is ready for use, neither does it block till that point.

cc XilunWu H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot merge
